### PR TITLE
Blend mode support

### PR DESCRIPTION
**Pull request by k-o-d-e-n**

### CODE REVIEW

**Comment by calda** [comment]
Thanks for the contribution @k-o-d-e-n! I'm going to convert this to a draft for now since CI isn't passing, but feel free to continue working on this. Once CI passes, I think we can merge this.
 - I think the build is probably failing because the new `BlendMode+Filter.swift` file isn't referenced in the xcode project. 
 - You'll need to run the snapshot test suite to create snapshot images for the sample asset you added

File: Sources/Private/Utility/Extensions/BlendMode+Filter.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
+//
+//  File.swift
+//  
+//
+//  Created by Denis Koryttsev on 10.05.2022.
+//
+
+extension BlendMode {
+    var filterName: String? {
```
[the remark this answers is not in the document]
Comment (k-o-d-e-n) [reply Sources/Private/Utility/Extensions/BlendMode+Filter.swift:9]: From [here](https://stackoverflow.com/a/42962239/7232781).

File: Sources/Private/CoreAnimation/Layers/BaseCompositionLayer.swift [hunk 2]
```diff
@@ -15,6 +15,7 @@ class BaseCompositionLayer: BaseAnimationLayer {
     super.init()
 
     setupSublayers()
+    compositingFilter = layerModel.blendMode.filterName
```
[the remark this answers is not in the document]
Comment (k-o-d-e-n) [reply Sources/Private/CoreAnimation/Layers/BaseCompositionLayer.swift:18]: The documentation is not always complete. I don't know, but it works. I compared with web, all the same.

[the remark this answers is not in the document]
Comment (calda) [reply Sources/Private/CoreAnimation/Layers/BaseCompositionLayer.swift:18]: good to know!

File: Sources/Private/Utility/Extensions/BlendMode+Filter.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (calda) [reply Sources/Private/Utility/Extensions/BlendMode+Filter.swift:9]:
```suggestion
    /// The Core Image filter name for this `BlendMode`, that can be applied to a `CALayer`'s `compositingFilter`.
    /// Supported compositing filters are defined here: https://developer.apple.com/library/archive/documentation/GraphicsImaging/Reference/CoreImageFilterReference/index.html#//apple_ref/doc/uid/TP30000136-SW71
    var filterName: String? {
```

File: Tests/Samples/blend_mode_test.json [hunk 3]
```diff
@@ -0,0 +1 @@
...,"refId":"precomp_6","w":100,"h":130},{"nm":"Precomp layer","ind":4,"ty":0,"ddd":0,"ip":0.000000,"op":1.000000,"st":0.000000,"ks":{"a":{"k":[0.000000,0.000000,0.000000]},"p":{"k":[187.869433,-0.166732,0.000000]},"s":{"k":[100.000000,100.000000,100.000000]},"r":{"k":0.000000},"o":{"k":100.000000}},"bm":0,"hasMask":false,"sr":1.000000,"tt":0,"hd":false,"refId":"precomp_4","w":100,"h":130},{"nm":"Precomp layer","ind":5,"ty":0,"ddd":0,"ip":0.000000,"op":1.000000,"st":0.000000,"ks":{"a":{"k":[0.000000,0.000000,0.000000]},"p":{"k":[93.608372,-0.364205,0.000000]},"s":{"k":[100.000000,100.000000,100.000000]},"r":{"k":0.000000},"o":{"k":100.000000}},"bm":0,"hasMask":false,"sr":1.000000,"tt":0,"hd":false,"refId":"precomp_2","w":100,"h":130},{"nm":"Precomp layer","ind":6,"ty":0,"ddd":0,"ip":0.000000,"op":1.000000,"st":0.000000,"ks":{"a":{"k":[0.000000,0.000000,0.000000]},"p":{"k":[0.000000,0.000000,0.000000]},"s":{"k":[100.000000,100.000000,100.000000]},"r":{"k":0.000000},"o":{"k":100.000000}},"bm":0,"hasMask":false,"sr":1.000000,"tt":0,"hd":false,"refId":"precomp_1","w":100,"h":130}],"fonts":{"list":[{"fName":"Helvetica-Regular","fFamily":"Helvetica","fStyle":"Regular","ascent":24.000000}]}}
```
Comment (calda) [inline Tests/Samples/blend_mode_test.json:1]: Could you move this from `Tests/Samples` to `Tests/Samples/Nonanimating/blend_mode_test.json` or `Tests/Samples/Issues/pr_1585.json`?

**Review by calda — APPROVED** [review]

File: Sources/Private/Utility/Extensions/BlendMode+Filter.swift [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        case .darken: return "darkenBlendMode"
+        case .lighten: return "lightenBlendMode"
+        case .colorDodge: return "colorDodgeBlendMode"
+        case .colorBurn: return "colorBurnBlendMode"
+        case .hardLight: return "hardLightBlendMode"
+        case .softLight: return "softLightBlendMode"
+        case .difference: return "differenceBlendMode"
+        case .exclusion: return "exclusionBlendMode"
+        case .hue: return "hueBlendMode"
+        case .saturation: return "multiplyBlendMode"
```
Comment (calda) [inline Sources/Private/Utility/Extensions/BlendMode+Filter.swift:26]: Should this be:

```suggestion
        case .saturation: return "saturationBlendMode"
```

File: Tests/Samples/blend_mode_test.json (hunk 3 above)
Comment (k-o-d-e-n) [reply Tests/Samples/blend_mode_test.json:1]: Previous snapshots need to be removed?

**Comment by calda** [comment]
@k-o-d-e-n, some of the checks are failing. I think you need to:
 - regenerate the snapshot images for the samples you added, to resolve the `Snapshot "Nonanimating/blend_mode_test (0%)" does not match reference` error. You can do that by either deleting the file on disk and having it be recreated on the next run, or by adding `isRecording = true` to the `testCoreAnimationEngine()` test case and rerunning the test
 - run the Swift liner (`bundle exec rake format:swift`)

**Comment by calda** [comment]
If the snapshot test case keeps failing, then the rendered output must be slightly nondeterministic. You an work around this type of issue by decreasing the comparison precision (example: https://github.com/airbnb/lottie-ios/blob/master/Tests/SnapshotConfiguration.swift#L43)

**Comment by k-o-d-e-n** [comment]
I have no failed tests with `blend_mode_test.json`. I don't know why it fails on CI.
Screenshot 2022-05-25 at 13 27 27

**Comment by calda** [comment]
I pushed some commits that address the CI failures.
 - For the `Typeface/B` snapshots, I regenerated them using `isRecording = true`. Looks like that animation has a blend mode setup so its snapshots were affected by this PR.
 - For the `blend_mode_test` snapshots, it looks like those were being rendered nondeterministically based on the environment running the tests (we've hit this a few times before). I fixed this by decreasing the comparison precision for that test to `0.99`
Will merge once checks pass. Thanks again for your contribution!

**Comment by k-o-d-e-n** [comment]
Thanks for help!

**Pull request merged**